### ISSUE TITLE
"Certain ARIA roles must contain particular children" reported for correct accessibility tree elements

### ISSUE THREAD

**Issue by dtopalov** [dequelabs/axe-core#4048]
### Product

axe-core

### Product Version

4.7.1

### Latest Version

- [X] I have tested the issue with the latest version of the product

#### Expectation
Do not report "aria-required-children" violation for correctly built accessibility trees when relationship between certain elements is achieved through aria attributes (the accessibility tree differs from the DOM tree).

#### Actual
Reports "aria-required-children" error for elements that are "owned" by a table row element from a different table.

#### How to Reproduce
Please consider the following simplified example (run in separate window and run the developer tools AXE extension in Chrome):

https://stackblitz.com/edit/js-wbinbi?file=index.js,index.html

It triggers the same report:

<img width="1185" alt="image" src="https://github.com/dequelabs/axe-core/assets/7533599/f0229499-5709-4ecd-8928-b8ab575cc702">

But the accessibility tree is correct:

<img width="531" alt="image" src="https://github.com/dequelabs/axe-core/assets/7533599/73c0f833-d614-44e2-9cf3-57f1e16665b6">

#### Additional context
We have a PivotGrid component, consisting of three separate tables (column headers, row headers, data), and an empty "cell" in the top-left corner between the row and column header tables:

<img width="1185" alt="image" src="https://github.com/dequelabs/axe-core/assets/7533599/a54dd7d7-2afd-4382-9e6d-d12f987cf7e6">

Our problem is that AXE reports the following problem:

[Ensures elements with an ARIA role that require child roles contain them](https://dequeuniversity.com/rules/axe/4.7/aria-required-children?application=AxeChrome) because the Empty cell needs a "columnheader" role, but is a direct child (in the DOM) of the role="grid" element, and all data cells have role "gridcell", but are also considered direct children of the role="grid" wrapper (even though all mentioned cells are referred through aria-owns by the respective rows that they need to be part of.

The accessibility tree we build seems correct, and assistive technologies announce the relations as expected, but the AXE error is still reported.

Can you please check whether the AXE tool evaluates connections between elements, built with aria attributes like role="none/presentation", aria-owns, etc., correctly?

**Comment by WilcoFiers** [comment]
In theory this grid is correct. In practice though, because you're relying on `aria-owns` for this, Safari isn't going to understand this grid. Unfortunately [`aria-owns`](https://a11ysupport.io/tech/aria/aria-owns_attribute) isn't supported in Safari at all. It's okay to use as a way to enhance the DOM tree, and axe lets you do that. But it's not something that can be relied on.

I think axe needs to make this problem clearer. Right now it's basically ignoring aria-owns. It should explicitly call  out the problem in the remediation message instead.

**Comment by veselints** [comment]
@WilcoFiers As far as I can see, axe-core properly interprets structure that uses `aria-owns` for building grids. Here is one highly complex example, where one `grid` component is created using 4 separate `<table>` elements:
https://runner.telerik.io/fullscreen/ITAhIsuJ/2

You will notice that no `aria-required-children` violation will be reported by axe.

If we leave aside the decision of Apple to disregard the living WAI-ARIA spec (it is their decision to deprive their own customers), the accessibility tree in both above cases is (technically) properly built. The question is, why axe reports a violation in one of the cases and not in the other? In my opinion both Grids are (technically) correct, so no error should be flagged.

**Comment by WilcoFiers** [comment]
Rereading the latest response. @veselints can you clarify this:

> The question is, why axe reports a violation in one of the cases and not in the other? 

What two cases are you seeing an inconsistency between?

**Comment by veselints** [comment]
Hi @WilcoFiers

Here is one example where axe reports "aria-required-children" error:
https://stackblitz.com/edit/js-wbinbi?file=index.js,index.html

Here is another example where axe does not identify any issues:
https://runner.telerik.io/fullscreen/ITAhIsuJ/2

In both cases, the accessibility tree forming the Grid is built using the `aria-owns` attribute. The tree has been properly constructed (grid>rowgroup>row>columnheader/gridcell). Hence, there should be no errors reported in both the above cases.

**Comment by straker** [comment]
@veselints For the first page (stackbliz.com) the table is not properly constructed and is reporting violations correctly. The top table is `role=grid` but then the very first child is a `role=columnheader`. `columnheader` requires having a parent `row` so the parent of `grid` is not allowed. Also the last nested table uses `role=presentation` with children `tr[role=none]` so the `td[role=gridcell]` elements are also not contained within a parent with `role=row`.

**Comment by straker** [comment]
Further context. Because the direct child of the table is owned by another element, the ARIA spec does not state which parent takes precedence. We have [an open issue about this](https://github.com/dequelabs/axe-core/issues/926) where we tried to figure it out.

**Comment by lifeiscontent** [comment]
Hitting the same `aria-required-children` false positive, but for menus instead of grids. Adding it here since it's the same `aria-owns` accessibility-tree-vs-DOM root cause (and the precedence question in #926).

floating-ui's `FloatingPortal` (used by Base UI and a lot of menu libraries) renders a visually hidden `<span aria-owns="{submenu-id}">` next to a submenu trigger so the portaled submenu lands in the right spot in the accessibility tree. When the trigger is a `menuitem` inside a `role="menu"`/`menubar`, that owner span becomes a DOM child of the menu, and axe reports:

> Element has children which are not allowed: span[aria-owns]

Minimal repro:

```html
<div role="menu">
  <button role="menuitem">Cut</button>
  <button role="menuitem" aria-haspopup="menu" aria-expanded="true" aria-controls="submenu">Share</button>
  <!-- floating-ui's owner span: reparents the portaled submenu in the a11y tree -->
  <span aria-owns="submenu"></span>
</div>
<div id="submenu" role="menu">
  <button role="menuitem">Copy link</button>
</div>
```

The span has no role; it's purely an `aria-owns` host. In the accessibility tree it isn't a meaningful child of the menu, but the rule counts it as a disallowed DOM child, same as the grid case above: the DOM tree and the a11y tree diverge because of `aria-owns`, and the rule reads the DOM tree.

Two questions for the team, since this affects basically every floating-ui/Base UI menu with a portaled submenu in non-modal mode:
1. Should a roleless `aria-owns` owner element be exempt from `aria-required-children`? It owns content elsewhere; it isn't itself content.
2. Is this the same precedence decision tracked in #926, and is there a direction there?

Happy to split this into its own issue if you'd rather keep this one grid-scoped. Upstream context: floating-ui/floating-ui#3424.

**Comment by WilcoFiers** [comment]
I think we're going to need to treat this as a false positive now. Support for `aria-owns` has developed far enough.

@lifeiscontent Your example does have a problem with it. While regular span elements get ignored in the accessibility tree, if you put a global property like `aria-owns` on them they no longer will. That means the structure of your

```
menu
- menuitem: Cut
- menuitem: Share
- generic
    -  menu
        - menuitem: Copy link
```

As a result, VoiceOver announces your top menu as having 3 items instead of 2. That's potentially a confusing user experience. There isn't really a way around this I think  Generics with aria-owns, are added to the accessibility tree, as are generics that are targeted by aria-owns. The only way this can work is if you put the menu in, and have it reference all the items in your portalled element:

```html
<div role="menu">
  <button role="menuitem">Cut</button>
  <button role="menuitem" aria-haspopup="menu" aria-expanded="true" aria-controls="submenu">Share</button>
  <!-- floating-ui's owner span: reparents the portaled submenu in the a11y tree -->
  <span role="menu" aria-owns="submenu-item1 submenu-item2"></span>
</div>
<div id="submenu">
  <button role="menuitem" id="submenu-item1">Copy link</button>
  <button role="menuitem" id="submenu-item2">Share on X</button>
</div>
```
